File: backend/internal/domain/movement.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

type MovementType string

const (
	MovementTypeIn         MovementType = "IN"
	MovementTypeOut        MovementType = "OUT"
	MovementTypeAdjustment MovementType = "ADJUSTMENT"
)

type StockMovement struct {
	ID            uuid.UUID    `json:"id" db:"id"`
	ItemID        uuid.UUID    `json:"itemId" db:"item_id"`
	MovementType  MovementType `json:"movementType" db:"movement_type" validate:"required"`
	Quantity      int          `json:"quantity" db:"quantity" validate:"required"`
	PreviousStock int          `json:"previousStock" db:"previous_stock"`
	NewStock      int          `json:"newStock" db:"new_stock"`
	Reference     *string      `json:"reference" db:"reference"`
	Notes         *string      `json:"notes" db:"notes"`
	CreatedBy     uuid.UUID    `json:"createdBy" db:"created_by"`
	CreatedAt     time.Time    `json:"createdAt" db:"created_at"`

	// Joined fields
	Item *Item `json:"item,omitempty"`
}

type CreateMovementRequest struct {
	ItemID       uuid.UUID    `json:"itemId" validate:"required"`
	MovementType MovementType `json:"movementType" validate:"required"`
	Quantity     int          `json:"quantity" validate:"required"`
	Reference    *string      `json:"reference"`
	Notes        *string      `json:"notes"`
}

File: mobile/lib/models/movement.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'item.dart';
import 'user.dart';

part 'movement.freezed.dart';
part 'movement.g.dart';

enum MovementType {
  @JsonValue('IN')
  stockIn,
  @JsonValue('OUT')
  stockOut,
  @JsonValue('ADJUSTMENT')
  adjustment,
}

@freezed
class StockMovement with _$StockMovement {
  const factory StockMovement({
    required String id,
    required String itemId,
    required MovementType movementType,
    required double quantity,
    required double previousStock,
    required double newStock,
    String? reference,
    String? notes,
    required String createdBy,
    required String createdAt,
    Item? item,
    User? user,
  }) = _StockMovement;

  factory StockMovement.fromJson(Map<String, dynamic> json) =>
      _$StockMovementFromJson(json);
}

@freezed
class CreateMovementRequest with _$CreateMovementRequest {
  const factory CreateMovementRequest({
    required String itemId,
    required MovementType movementType,
    required double quantity,
    String? reference,
    String? notes,
  }) = _CreateMovementRequest;

  factory CreateMovementRequest.fromJson(Map<String, dynamic> json) =>
      _$CreateMovementRequestFromJson(json);
}
